crates/kubelet/src/state.rs

```diff
@@ -203,26 +216,42 @@ impl<PodState> Transition<PodState> {
     /// }
     /// ```
     #[allow(clippy::boxed_local)]
-    pub fn next<I: State<PodState>, S: State<PodState>>(_i: Box<I>, s: S) -> Transition<PodState>
+    pub fn next<I: State<ProviderState, PodState>, S: State<ProviderState, PodState>>(
+        _i: Box<I>,
+        s: S,
+    ) -> Transition<ProviderState, PodState>
     where
         I: TransitionTo<S>,
     {
         Transition::Next(StateHolder { state: Box::new(s) })
     }
+
+    #[allow(clippy::boxed_local)]
+    pub fn next_unchecked<S: State<ProviderState, PodState>>(
+        s: S,
+    ) -> Transition<ProviderState, PodState> {
+        Transition::Next(StateHolder { state: Box::new(s) })
+    }
 }
 
 #[async_trait::async_trait]
 /// Allow for asynchronous cleanup up of PodState.
 pub trait AsyncDrop: Sized {
+    type ProviderState;
     /// Clean up PodState.
-    async fn async_drop(self);
+    async fn async_drop(self, provider_state: &mut Self::ProviderState, pod: &Pod);
 }
 
 #[async_trait::async_trait]
 /// A trait representing a node in the state graph.
-pub trait State<PodState>: Sync + Send + 'static + std::fmt::Debug {
+pub trait State<ProviderState, PodState>: Sync + Send + 'static + std::fmt::Debug {
     /// Provider supplies method to be executed when in this state.
-    async fn next(self: Box<Self>, pod_state: &mut PodState, pod: &Pod) -> Transition<PodState>;
+    async fn next(
+        self: Box<Self>,
+        provider_state: std::sync::Arc<tokio::sync::RwLock<ProviderState>>,
```

Comment:
But in practice that means that providers just shunt the shared state into `PodState::shared`.  I agree with the desire, but I actually feel that making that shared state explicit could provide us with better opportunities to specify and enforce safe ways of updating it rather than requiring every provider to implement their own.  Of course, needless to say I haven't done that here - and yes, the best way to present this so that it was safe was one of the things I was most unsure about, so very keen to have this discussion and find a better way than my sighting shot!

Reply:
This is true. I wanted it to be something that developers were forced to have ownership of (concurrency decisions). But it probably doesnt make a meaningful difference.